packages/astro/src/i18n/index.ts

```diff
@@ -1,21 +1,56 @@
+import { AstroError } from '../core/errors/index.js';
+import { MissingLocale } from '../core/errors/errors-data.js';
 import type { AstroConfig } from '../@types/astro.js';
-import type { Logger } from '../core/logger/core.js';
+import { shouldAppendForwardSlash } from '../core/build/util.js';
 
+type GetI18nBaseUrl = {
+	locale: string;
+	base: string;
+	locales: string[];
+	trailingSlash: AstroConfig['trailingSlash'];
+	format: AstroConfig['build']['format'];
+};
 /**
  * The base URL
  */
-export function getI18nBaseUrl(locale: string, config: AstroConfig, logger: Logger) {
-	const base = config.base;
-
-	if (!config.experimental.i18n) {
-		logger.error('i18n', "The project isn't using i18n features, no need to use this function.");
-		return base ? `/${base}/` : '/';
+export function getI18nBaseUrl({ locale, base, locales, trailingSlash, format }: GetI18nBaseUrl) {
+	if (!locale.includes(locale)) {
+		throw new AstroError({
+			...MissingLocale,
+			message: MissingLocale.message(locale, locales),
+		});
 	}
 
-	if (base) {
-		logger.debug('i18n', 'The project has a base directory, using it.');
-		return `${base}/${locale}/`;
+	const normalizedLocale = normalizeLocale(locale);
+	if (shouldAppendForwardSlash(trailingSlash, format)) {
+		return `${base}${normalizedLocale}/`;
 	} else {
-		return `/${locale}/`;
+		return `${base}/${normalizedLocale}`;
 	}
 }
+
+export function getLocalesBaseUrl(
+	base: string,
+	locales: string[],
+	trailingSlash: AstroConfig['trailingSlash'],
+	format: AstroConfig['build']['format']
```

Comment:
I think it should be using an object and be extracted to a type like `GetI18nBaseUrl`?